// ollama/dist/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
import * as utils from "./utils.js";
export class Ollama {
    constructor(config) {
        var _a;
        this.config = {
            address: (_a = config === null || config === void 0 ? void 0 : config.address) !== null && _a !== void 0 ? _a : "http://localhost:11434"
        };
    }
    tags() {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield utils.get(`${this.config.address}/api/tags`);
            const json = yield response.json();
            return json.models.map(m => ({
                name: m.name,
                modifiedAt: new Date(m.modified_at),
                size: m.size
            }));
        });
    }
    generate(model, prompt) {
        return __asyncGenerator(this, arguments, function* generate_1() {
            var _a, e_1, _b, _c;
            const response = yield __await(utils.post(`${this.config.address}/api/generate`, { model, prompt }));
            if (!response.body) {
                throw new Error("Missing body");
            }
            try {
                for (var _d = true, _e = __asyncValues(response.body), _f; _f = yield __await(_e.next()), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const chunk = _c;
                        const messages = chunk.toString().split("\n").filter(s => s.length !== 0);
                        for (const message of messages) {
                            const res = JSON.parse(message);
                            if (res.done) {
                                return yield __await({
                                    model: res.model,
                                    createdAt: new Date(res.created_at),
                                    context: res.context,
                                    totalDuration: res.total_duration,
                                    loadDuration: res.load_duration,
                                    promptEvalCount: res.prompt_eval_count,
                                    evalCount: res.eval_count,
                                    evalDuration: res.eval_duration
                                });
                            }
                            yield yield __await(res.response);
                        }
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
            throw new Error("Did not recieve done response in stream.");
        });
    }
    create(name, path) {
        return __asyncGenerator(this, arguments, function* create_1() {
            var _a, e_2, _b, _c;
            const response = yield __await(utils.post(`${this.config.address}/api/create`, { name, path }));
            if (!response.body) {
                throw new Error("Missing body");
            }
            try {
                for (var _d = true, _e = __asyncValues(response.body), _f; _f = yield __await(_e.next()), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const chunk = _c;
                        const messages = chunk.toString().split("\n").filter(s => s.length !== 0);
                        for (const message of messages) {
                            const res = JSON.parse(message);
                            yield yield __await(res.status);
                        }
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_2) throw e_2.error; }
            }
        });
    }
    copy(source, destination) {
        return __awaiter(this, void 0, void 0, function* () {
            yield utils.post(`${this.config.address}/api/copy`, {
                source,
                destination
            });
        });
    }
    delete(name) {
        return __awaiter(this, void 0, void 0, function* () {
            yield utils.del(`${this.config.address}/api/delete`, { name });
        });
    }
    pull(name) {
        var _a, _b, _c;
        return __asyncGenerator(this, arguments, function* pull_1() {
            var _d, e_3, _e, _f;
            const response = yield __await(utils.post(`${this.config.address}/api/pull`, { name }));
            if (!response.body) {
                throw new Error("Missing body");
            }
            try {
                for (var _g = true, _h = __asyncValues(response.body), _j; _j = yield __await(_h.next()), _d = _j.done, !_d;) {
                    _f = _j.value;
                    _g = false;
                    try {
                        const chunk = _f;
                        const messages = chunk.toString().split("\n").filter(s => s.length !== 0);
                        for (const message of messages) {
                            const res = JSON.parse(message);
                            yield yield __await({
                                status: res.status,
                                digest: (_a = res["digest"]) !== null && _a !== void 0 ? _a : "",
                                total: (_b = res["total"]) !== null && _b !== void 0 ? _b : 0,
                                completed: (_c = res["completed"]) !== null && _c !== void 0 ? _c : 0
                            });
                        }
                    }
                    finally {
                        _g = true;
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (!_g && !_d && (_e = _h.return)) yield __await(_e.call(_h));
                }
                finally { if (e_3) throw e_3.error; }
            }
        });
    }
    embeddings(model, prompt) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield utils.post(`${this.config.address}/api/embeddings`, {
                model,
                prompt
            });
            const json = yield response.json();
            return json.embedding;
        });
    }
}
